### PR TITLE
Separating fpg and apriori code, cleanup

### PR DESCRIPTION
**Pull request by edgan8**
Adding new explanation result classes that keep track of
raw counts for both interpretability and extensibility. This involved
separating the fpg and apriori dependencies, in the future we can merge
fpgrowth to use the new code.

Compare ItemsetResult with ExplanationResult, note how ExplanationResult is easier to work with since it tracks all the raw counts.

### CODE REVIEW

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package edu.stanford.futuredata.macrobase.analysis.summary.apriori;
+
+
+import java.util.Map;
+import java.util.StringJoiner;
+
+public class ExplanationResult {
+    private Map<String, String> matcher;
+    private double matchedOutlier, matchedCount;
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java]: Having these raw counts makes things much easier for users to interpret (How many points is this? vs What is a "support" fraction?) Also lets us add new measures easily since we have these raw counts.

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            double totalOutlier,
+            double totalCount
+    ) {
+        this.matcher = matcher;
+        this.matchedOutlier = matchedOutlier;
+        this.matchedCount = matchedCount;
+        this.totalOutlier = totalOutlier;
+        this.totalCount = totalCount;
+
+        this.rateLift = (matchedOutlier / matchedCount) / (totalOutlier / totalCount);
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java]: This isn't the risk ratio: is the ratio of current outlier rate to global outlier rate. It's easier to compute, more well behaved, and plays better with the new pruning rules. 

Under risk ratio, if a "result" trivially matched a huge population including all the outliers it would have infinite risk ratio, masking more meaningful results.

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        return support;
+    }
+
+    public String prettyPrint() {
+        StringJoiner joiner = new StringJoiner(",");
+        matcher.forEach((k, v) -> joiner.add(k+"="+v));
+
+        return String.format(
+                "--\n" +
+                        "outlier rate: %d / %d\n" +
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java]: This shows the rate of outliers matched / total matched for the result. I found it more useful for understanding the results than just looking at support and RR

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/fpg/FPGExplanation.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represents a summarization result, which contains a list of attribute values
+ * and other statistics about the underlying process, e.g. num of tuples observed
+ * so far.
+ */
+public class FPGExplanation implements Explanation {
```
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/fpg/FPGExplanation.java]: Kept around for backwards compatibility until we move fpgrowth to use the new explanations in a future PR

File: core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java [hunk 5]
```diff
@@ -76,6 +78,15 @@ public BatchSummarizer getSummarizer(String outlierColumnName) throws MacrobaseE
                 summarizer.setMinRiskRatio(minRiskRatio);
                 return summarizer;
             }
+            case "fpgrowth": {
```
Comment (pbailis) [inline core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java]: prefer enums here, same as with @fabuzaid21... do you think it's better to use strings?

File: core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java [hunk 6]
```diff
@@ -76,6 +78,15 @@ public BatchSummarizer getSummarizer(String outlierColumnName) throws MacrobaseE
...
+                FPGrowthSummarizer summarizer = new FPGrowthSummarizer();
+                summarizer.setOutlierColumn(outlierColumnName);
+                summarizer.setAttributes(attributes);
+                summarizer.setMinSupport(minSupport);
+                summarizer.setMinRiskRatio(minRiskRatio);
+                summarizer.setUseAttributeCombinations(true);
+                return summarizer;
+            }
```
Comment (pbailis) [inline core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java]: Feature request: for apriori, can we add a config option to set the maximum order?

Comment (pbailis) [inline core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java]: @fabuzaid21 needed this last night, should be easy to add and we don't want to make people edit the code and recompile :)

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java [hunk 7]
```diff
@@ -1,111 +1,7 @@
...
-    @Override
-    public String toString() {
-        return "Explanation{" +
-                "numOutliers=" + numOutliers +
-                ", numInliers=" + numInliers +
-                ", itemsets=" + itemsets +
-                '}';
-    }
+public interface Explanation {
+    String prettyPrint();
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java]: is this a common idiom for Java libraries?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APExplanation.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * An explanation attempts to summarize the differences between the outliers and inliers.
+ * It is composed of multiple ExplanationResult results, each of which has a matcher which
+ * is an atomic classifier for rows in a dataset.
+ */
+public class APExplanation implements Explanation {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APExplanation.java]: What is "AP"? Why not just make it Apriori?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APExplanation.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            List<ExplanationResult> results,
+            double numOutliers,
+            double numTotal
+            ) {
+        this.results = new ArrayList<>(results);
+        this.numOutliers = numOutliers;
+        this.numTotal = numTotal;
+    }
+
+    public void sortBySupport() {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APExplanation.java]: Not clear why this is necessary and/or why it's not done by default.

Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APExplanation.java]: Basically, could do this at construction time, and then have clients resort as needed.

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java [hunk 10]
```diff
@@ -313,8 +311,8 @@ private void countSingles(List<int[]> encoded, double[] countCol, double[] outli
     }
 
     @Override
-    public Explanation getResults() {
-        List<AttributeSet> results = new ArrayList<>();
+    public APExplanation getResults() {
+        List<ExplanationResult> results = new ArrayList<>();
         for (int o = 1; o <= 3; o++) {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java]: make this configurable

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java [hunk 11]
```diff
@@ -324,28 +322,32 @@ public Explanation getResults() {
...
-                numEvents - numOutliers,
                 numOutliers,
-                timings[1]+timings[2]+timings[3]
+                numEvents
         );
         finalExplanation.sortBySupport();
         return finalExplanation;
     }
 
+    private double computeRatio(
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java]: comment to describe what this statistic is computing

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java (hunk 1 above)
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java]: Wfm

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java (hunk 2 above)
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResult.java]: How does this relate to computeRatio?

Do we also want to expose riskRatio?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResultGrouper.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
+package edu.stanford.futuredata.macrobase.analysis.summary.apriori;
+
+import edu.stanford.futuredata.macrobase.datamodel.DataFrame;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class ExplanationResultGrouper {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResultGrouper.java]: Comment to explain why;

Also, kind of annoying this only works for APExplanations

Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResultGrouper.java]: Is there a more general interface this could take?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResultGrouper.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+    public ExplanationResultGrouper(
+            APExplanation e,
+            DataFrame d
+    ) {
+        this.explanation = e;
+        this.data = d;
+    }
+
+    public ArrayList<Integer> getMatches(Map<String, String> matcher) {
```
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/ExplanationResultGrouper.java]: needs comment, especially why returning Integer; also returning Arraylist vs map...

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/fpg/FPGExplanation.java (hunk 4 above)
Comment (pbailis) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/fpg/FPGExplanation.java]: fine

File: core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java (hunk 6 above)
Comment (fabuzaid21) [inline core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/BasicBatchPipeline.java]: I was actually about to submit another PR to set a maximum order LOL. @edgan8 do you wanna incorporate it in this PR, or should I push mine?

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java (hunk 10 above)
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/apriori/APrioriSummarizer.java]: You can choose which ratio you want now, it's defined when you create a summarizer object

File: lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java (hunk 7 above)
Comment (edgan8) [inline lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/Explanation.java]: Which part are you referring to? Defining interfaces like this is a pretty common way to support both old and new code.

**Comment by edgan8** [comment]
I've kept the string parameters for now: since they are only used to create a single object in the toplevel pipeline it doesn't seem worth creating enums to clutter things up. We should definitely switch to enums if these strings start showing up more than once.

File: core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/CubePipeline.java [hunk 14]
```diff
@@ -76,7 +75,7 @@ public Explanation results() throws Exception {
         summarizer.setCountColumn(classifier.getCountColumnName());
         summarizer.setAttributes(attributes);
         summarizer.setMinSupport(minSupport);
-        summarizer.setMinRiskRatio(minRiskRatio);
+        summarizer.setMinRatioMetric(minRatioMetric);
```
Comment (edgan8) [inline core/src/main/java/edu/stanford/futuredata/macrobase/pipeline/CubePipeline.java]: This sets the kind of "ratio metric" we want to use: can be risk ratio or anything lse

**Comment by fabuzaid21** [comment]
LGTM
